### ISSUE TITLE
JSON PATCH support

### ISSUE THREAD

**Issue by dkarlovi** [api-platform/core#759]
#147 suggested adding a PATCH support for partials updates, it was closed as PUT already does that.

How about adding a [JSON PATCH](http://jsonpatch.com/) support? Difference here is you're able to do operations WITH entities, not only ON entites, quoted from [Why PATCH is Good for Your HTTP API](https://www.mnot.net/blog/2012/09/05/patch)

> PATCH is atomic, which means that if you need to do some complex changes to the state of a resource, you can do it with confidence. It also means that you can create synthetic resources and PATCH them if you need to orchestrate changes to the state of several resources.

**Comment by dunglas** [comment]
It would definitely be a great improvement to have support for JSON PATCH.

**Comment by dunglas** [comment]
On the other end, [JSON Merge Patch](https://tools.ietf.org/html/rfc7386) will be easier to implement, and easier to use client-side.

**Comment by teohhanhui** [comment]
http://erosb.github.io/post/json-patch-vs-merge-patch/

**Comment by asimonf** [comment]
@dunglas given that the content-type of both methods is different, couldn't both be implemented? I understand that Merge Patch has limitations, but for many cases, it's simplicity could be great for certain applications.

What I mean is, I don't see why we should have to pick one or the other. They are just different formats and both could be supported. If that's ok with you, I might have a go at trying to implement merge patch, time permitting.

**Comment by dunglas** [comment]
Yew we can support both. It should be possible to support JSON MERGE PATCH with just a few modifications to the current system.

**Comment by teohhanhui** [comment]
Implementing JSON Patch will force us to improve our design. And it will allow us to deprecate the current partial PUT semantics (which is wrong). :smile:

**Comment by asimonf** [comment]
@teohhanhui I don't see why both (JSON MERGE and JSON PATCH) can't be implemented. I don't disagree with you and I'm sure if a PR appears for JSON PATCH, @dunglas would be willing to merge assuming it passes review. In my use-case, the simplicity of JSON MERGE trumps the flexibility of JSON PATCH.

**Comment by teohhanhui** [comment]
I'm okay with having both. :)

**Comment by teohhanhui** [comment]
Also, I think we can deprecate the partial PUT once we have JSON Patch.
Because it already covers the existing semantics.

**Comment by StephenOTT** [comment]
anyone actively working on this?

**Comment by dunglas** [comment]
AFAIK, no. But we should probably consider using JSON Merge Patch instead: https://tools.ietf.org/html/rfc7386

**Comment by StephenOTT** [comment]
@dunglas can this issue be closed in favour of https://github.com/api-platform/core/pull/1175 ?

**Comment by teohhanhui** [comment]
[JSON API](http://jsonapi.org) is only one media type. This is about PATCH support for all of the supported media types, and I'd think especially JSON-LD / Hydra.

**Comment by dkarlovi** [comment]
@dunglas IMO this shouldn't be closed in favor of JSON Merge since Merge is only a partial PUT.

JSON PATCH is a diff to a JSON document so, if a document represents a collection of your resources, you could:

1. append a new resource into a collection
2. remove a resource from a collection
3. re-order (some) collection items without touching (or mentioning) the rest

It solves the partial collection management problem well and IMO it should be kept as a future option.

**Comment by bendavies** [comment]
this seems a step too far at the moment. for the moment, why can we not just make `PUT` mean PUT` semantically, rather than `PATCH`? is is difficult?

**Comment by teohhanhui** [comment]
@bendavies Yes, but we need to make it opt-in for BC. And when we have proper JSON Patch support, we can deprecate not using the correct semantics.

**Comment by alanpoulain** [comment]
JSON Merge Patch support has landed in 2.5.

**Comment by teohhanhui** [comment]
Let's keep this open as we only support JSON Merge Patch, not JSON Patch.

**Comment by Muspi** [comment]
Hi, how do you manage partial Validation constraints with PATCH actions?
I'm facing some problems to achieve this, but I do not have solution.

# #1637

**Comment by DurandA** [comment]
I am attempting to implement a subset of JSON Patch/RFC 6902 in order to partially update collections. See discussion https://github.com/api-platform/core/discussions/4882.

**Comment by tobiasgv** [comment]
is there anything new/planned with regard to `application/json-patch+json` support? I keep coming across use cases where I need to do atomic PATCHes of resources with nested entities.

**Comment by rvanlaak** [comment]
Now Symfony's ObjectMapper (still experimental) is out, defining a DTO as `input` on an operation is sufficient in order to patch resources. The mapper will skip uninitialized properties, meaning that partial updates will work. 

I'm not fully sure whether there is documentation yet on a processor that was added to APIP 4.2, but I'm already using it in production on a couple of endpoints.
